internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,344 @@
+		entryCh:    make(chan servicelog.Entry),
+		pullers:    newPullerGroup(target.Name),
+		timer:      newTimer(),
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	g.tomb.Go(g.pullers.tomb.Wait)
+
+	return g, nil
+}
+
+func fillDefaultArgs(args logGathererArgs) logGathererArgs {
+	if args.bufferTimeout == 0 {
+		args.bufferTimeout = bufferTimeout
+	}
+	if args.timeoutFinalFlush == 0 {
+		args.timeoutFinalFlush = timeoutFinalFlush
+	}
+	if args.newClient == nil {
+		args.newClient = newLogClient
+	}
+	return args
+}
+
+// PlanChanged is called by the LogManager when the plan is changed, if this
+// gatherer's target exists in the new plan.
+func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	// Remove old pullers
+	for _, svcName := range g.pullers.Services() {
+		svc, svcExists := pl.Services[svcName]
+		if !svcExists {
+			g.pullers.Remove(svcName)
+			continue
+		}
+
+		tgt := pl.LogTargets[g.targetName]
+		if !svc.LogsTo(tgt) {
+			g.pullers.Remove(svcName)
+		}
+	}
+
+	// Add new pullers
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until ServiceStarted
+			continue
+		}
+
+		g.pullers.Add(service.Name, buffer, g.entryCh)
+	}
+}
+
+// ServiceStarted is called by the LogManager on the start of a service which
+// logs to this gatherer's target.
+func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullers.Add(service.Name, buffer, g.entryCh)
+}
+
+// The main control loop for the logGatherer. loop receives logs from the
+// pullers on entryCh, and writes them to the client. It also flushes the
+// client periodically, and exits when the gatherer's tomb is killed.
+func (g *logGatherer) loop() error {
+	defer g.timer.Stop()
+
+mainLoop:
+	for {
+		select {
+		case <-g.tomb.Dying():
+			break mainLoop
+
+		case <-g.timer.Expired():
+			// Mark timer as unset
+			g.timer.Stop()
+			g.handleClientErr(g.client.Flush(g.clientCtx))
+
+		case entry := <-g.entryCh:
+			g.handleClientErr(g.client.Write(g.clientCtx, entry))
+			g.timer.EnsureSet(g.bufferTimeout)
+		}
+	}
+
+	// Final flush to send any remaining logs buffered in the client
+	// We need to create a new context, as the previous one may have been cancelled.
+	ctx, cancel := context.WithTimeout(context.Background(), g.timeoutFinalFlush)
+	defer cancel()
+	g.handleClientErr(g.client.Flush(ctx))
+	return nil
+}
+
+func (g *logGatherer) handleClientErr(err error) {
+	if err == nil {
+		return
+	}
+
+	if e := (clienterr.Backoff{}); errors.As(err, &e) {
+		logger.Noticef("Target %q: too many requests, backing off", g.targetName)
+		g.timer.retryAfter = e.RetryAfter
+	}
+
+	if errors.As(err, &clienterr.Timeout{}) {
```

Comment:
Actually, why do we need a custom type for timeouts at all? If we didn't, it would just fall through and log below, something like `Cannot flush logs to target "foo": request timeout`.

Reply:
honestly I'm very confused about how to design custom errors that work with `errors.Is` and `errors.As`. It certainly doesn't seem intuitive. Maybe I'll ask Tom for some guidance here, he seems to know about this stuff

Reply:
@benhoyt we need custom error types so that clients can talk to the gatherer in a way it understands, and the gatherer can correctly respond. It saves us from having to understand potentially dozens of different error types in the gatherer - instead, clients are responsible for translating their errors into one of a few well-known types, which the gatherer knows how to deal with.

Maybe we don't currently *need* the `Timeout` type, as the gatherer is currently not doing anything in response to a timeout, but it seemed like a reasonable thing to add.

Reply:
Yep, I understand why we need custom error types in general (`Backoff`), but it's exactly because we're not using `Timeout` specially that (I think) we don't need it. Re "reasonable thing to add" -- I'd agree if this were a public package, but its in "internals", so we're free to change it at any time. In other words, we can just add it if and when we need it.

Reply:
Ok, well how should the gatherer react to a request timeout?

If the answer is "continue as usual", I'm happy to remove `Timeout`.

Reply:
I've now removed `Timeout` as it was unnecessary.